src/Http/Routing/src/RouteBase.cs

```diff
@@ -262,8 +262,8 @@ private void EnsureLoggers(HttpContext context)
             if (_logger == null)
             {
                 var factory = context.RequestServices.GetRequiredService<ILoggerFactory>();
-                _logger = factory.CreateLogger(typeof(RouteBase).FullName);
                 _constraintLogger = factory.CreateLogger(typeof(RouteConstraintMatcher).FullName);
+                _logger = factory.CreateLogger(typeof(RouteBase).FullName);
```

Comment:
Perhaps I'm wrong or misunderstanding how things work but it was my understanding that without `volatile`, it's entirely possible for lines like this to be reordered by the CPU. The only way to guarantee that every core/thread will see `_constraintLogger` *definitely assigned* before `_logger` is would be locking or using `volatile` fields.

If this seems to fix the race, then maybe it's moot, but I don't know that we can **assert** strongly that the ordering issue is resolved by a change like this.

Reply:
Re-ordering virtual method calls?

Reply:
If we *want* to be super clean and clear about this, I'd just slap a lock around the init, or use `LazyInitializer` with a valuetuple. 

I think I went through the same steps you did and then saw that these are iface calls. The compiler can't reason about that.

Reply:
I believe this should be enough. I don't think we mark things volatile almost anywhere in the stack. There are a bunch of volatile things in Kestrel, but id rather use https://docs.microsoft.com/en-us/dotnet/api/system.threading.volatile.write?view=netframework-4.8#System_Threading_Volatile_Write__1___0____0_ instead as this is a one-off. volatile has the potential to have perf implications

Reply:
I've updated it to use Volatile.Write cuz it's cheap

Reply:
> Re-ordering virtual method calls?

Maybe not, I'm really just being conservative here and don't entirely know what I'm talking about 😛.


> If we _want_ to be super clean and clear about this, I'd just slap a lock around the init, or use `LazyInitializer` with a valuetuple.

If this isn't a hot path, that seem safest.

> I've updated it to use Volatile.Write cuz it's cheap

🤷‍♂ 	probably ok?

Reply:
> If we want to be super clean and clear about this, I'd just slap a lock around the init, or use LazyInitializer with a valuetuple.

That would be better IMO.

cc @stephentoub PTAL

Reply:
You'd ideally also have a Volatile.Read on the read of _logger; that is, for example, what LazyInitializer.EnsureInitialized itself does, to avoid any concerns on weaker memory models around reordered, introduced, or coalesced reads.  I agree with @davidfowl, @anurse, and @rynowak that it'd be better to use a lock or Lazy or LazyInitializer; they exist in part to discourage devs from writing their own lock-free lazy initialization.

Reply:
Fair enough, I'll change this to use a LazyInitializer then.
@stephentoub thanks!